File: Camera_main_c/MacbethDetector.cs
using Emgu.CV;
using Emgu.CV.Structure;
using Emgu.CV.CvEnum;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text.Json;
using System.IO;
using Emgu.CV.Util;
using System.Linq;

public static class MacbethDetector
{
    private static Point[] OrderPoints(Point[] pts)
    {
        var rect = new Point[4];
        var sum = pts.Select(p => p.X + p.Y).ToArray();
        var diff = pts.Select(p => p.X - p.Y).ToArray();

        rect[0] = pts[Array.IndexOf(sum, sum.Min())];    // top-left
        rect[2] = pts[Array.IndexOf(sum, sum.Max())];    // bottom-right
        rect[1] = pts[Array.IndexOf(diff, diff.Min())];  // top-right
        rect[3] = pts[Array.IndexOf(diff, diff.Max())];  // bottom-left

        return rect;
    }

    public static (Mat WarpedFrame, Rectangle[] Squares) DetectMacbethInScene(Mat frameRaw, string cacheFile)
    {
        using var frameHsv = new Mat();
        CvInvoke.CvtColor(frameRaw, frameHsv, ColorConversion.Bgr2Hsv);

        // Détection du cadre noir
        var lowerBlack = new ScalarArray(new MCvScalar(0, 0, 0));
        var upperBlack = new ScalarArray(new MCvScalar(180, 100, 30));
        
        using var blackMask = new Mat();
        CvInvoke.InRange(frameHsv, lowerBlack, upperBlack, blackMask);

        // Nettoyage du masque
        var kernel = CvInvoke.GetStructuringElement(ElementShape.Rectangle, new Size(7, 7), new Point(-1, -1));
        CvInvoke.MorphologyEx(blackMask, blackMask, MorphOp.Close, kernel, new Point(-1, -1), 1, BorderType.Default, new MCvScalar());
        CvInvoke.MorphologyEx(blackMask, blackMask, MorphOp.Open, kernel, new Point(-1, -1), 1, BorderType.Default, new MCvScalar());

        using var contours = new VectorOfVectorOfPoint();
        CvInvoke.FindContours(blackMask, contours, null, RetrType.External, ChainApproxMethod.ChainApproxSimple);

        Point[] frameCorners = null;
        double maxArea = 0;

        for (int i = 0; i < contours.Size; i++)
        {
            var contour = contours[i];
            double area = CvInvoke.ContourArea(contour);
            
            if (area > 1000)
            {
                var peri = CvInvoke.ArcLength(contour, true);
                var approx = new VectorOfPoint();
                CvInvoke.ApproxPolyDP(contour, approx, 0.02 * peri, true);

                if (approx.Size == 4 && area > maxArea)
                {
                    maxArea = area;
                    frameCorners = approx.ToArray();
                }
            }
        }

        if (frameCorners == null)
        {
            throw new Exception("Aucun grand rectangle noir détecté.");
        }

        var orderedCorners = OrderPoints(frameCorners);
        
        // Calcul des dimensions de l'image redressée
        int targetWidth = (int)Math.Max(
            Distance(orderedCorners[1], orderedCorners[0]),
            Distance(orderedCorners[2], orderedCorners[3]));
        
        int targetHeight = (int)Math.Max(
            Distance(orderedCorners[3], orderedCorners[0]),
            Distance(orderedCorners[2], orderedCorners[1]));

        var dstPoints = new PointF[]
        {
            new PointF(0, 0),
            new PointF(targetWidth - 1, 0),
            new PointF(targetWidth - 1, targetHeight - 1),
            new PointF(0, targetHeight - 1)
        };

        var perspectiveMatrix = CvInvoke.GetPerspectiveTransform(
            orderedCorners.Select(p => new PointF(p.X, p.Y)).ToArray(),
            dstPoints);

        var frameWarped = new Mat();
        CvInvoke.WarpPerspective(frameRaw, frameWarped, perspectiveMatrix, new Size(targetWidth, targetHeight));

        // Détection des carrés de couleur
        var squares = DetectColorSquares(frameWarped);

        // Sauvegarde dans le cache
        SaveToCache(cacheFile, squares, frameWarped);

        return (frameWarped, squares);
    }

    private static double Distance(Point p1, Point p2)
    {
        return Math.Sqrt(Math.Pow(p2.X - p1.X, 2) + Math.Pow(p2.Y - p1.Y, 2));
    }

    public static Mat GetAverageColors(Mat frameRaw, string cacheFile, bool detectSquares)
    {
        Mat frameWarped;
        Rectangle[] squares;

        if (!detectSquares && File.Exists(cacheFile))
        {
            var cacheData = LoadFromCache(cacheFile);
            frameWarped = new Mat(cacheData.WarpedImagePath);
            squares = cacheData.Squares;
        }
        else
        {
            (frameWarped, squares) = DetectMacbethInScene(frameRaw, cacheFile);
        }

        var colors = new Mat(24, 1, DepthType.Cv8U, 3);
        
        for (int i = 0; i < squares.Length; i++)
        {
            var roi = new Mat(frameWarped, squares[i]);
            var mean = CvInvoke.Mean(roi);
            colors.SetValue(i, 0, new byte[] { (byte)mean.V0, (byte)mean.V1, (byte)mean.V2 });
        }

        return colors;
    }

    private static void SaveToCache(string cacheFile, Rectangle[] squares, Mat warpedFrame)
    {
        var warpedImagePath = cacheFile.Replace(".json", "_warped.png");
        CvInvoke.Imwrite(warpedImagePath, warpedFrame);

        var cacheData = new
        {
            Squares = squares,
            WarpedImagePath = warpedImagePath
        };

        File.WriteAllText(cacheFile, JsonSerializer.Serialize(cacheData));
    }

    private static (Rectangle[] Squares, string WarpedImagePath) LoadFromCache(string cacheFile)
    {
        var cacheContent = File.ReadAllText(cacheFile);
        var cacheData = JsonSerializer.Deserialize<dynamic>(cacheContent);
        return (
            cacheData.GetProperty("Squares").EnumerateArray()
                .Select(s => new Rectangle(
                    s.GetProperty("X").GetInt32(),
                    s.GetProperty("Y").GetInt32(),
                    s.GetProperty("Width").GetInt32(),
                    s.GetProperty("Height").GetInt32()
                )).ToArray(),
            cacheData.GetProperty("WarpedImagePath").GetString()
        );
    }

    private static Rectangle[] DetectColorSquares(Mat frameWarped)
    {
        // Implémentation de la détection des carrés de couleur
        var squares = new List<Rectangle>();
        
        // Paramètres de détection
        int squareSize = frameWarped.Width / 6;  // Approximation pour une charte 4x6
        int spacing = squareSize / 8;
        
        // Création de la grille 4x6
        for (int row = 0; row < 4; row++)
        {
            for (int col = 0; col < 6; col++)
            {
                squares.Add(new Rectangle(
                    col * (squareSize + spacing) + spacing,
                    row * (squareSize + spacing) + spacing,
                    squareSize,
                    squareSize
                ));
            }
        }
        
        return squares.ToArray();
    }

    private static void ProcessPoints(Point[] points)
    {
        // Convertir l'expression lambda en délégué explicite
        Func<Point, int> getX = p => p.X;
        var xCoords = points.Select(getX).ToArray();
        
        Func<Point, int> getY = p => p.Y;
        var yCoords = points.Select(getY).ToArray();
    }

    private static void SetMatrixValue(Mat matrix, byte[] values, int row)
    {
        unsafe
        {
            byte* ptr = (byte*)matrix.DataPointer;
            int offset = row * 3;
            ptr[offset] = values[0];
            ptr[offset + 1] = values[1];
            ptr[offset + 2] = values[2];
        }
    }
} 

File: Camera_POO/macbeth_detector.py
import cv2
import numpy as np
import json
import os
from config import config

class MacbethDetector:
    """Classe responsable de la détection et analyse de la charte Macbeth"""
    
    def __init__(self):
        self.cache_data = {}
        
    def _order_points(self, pts: np.ndarray) -> np.ndarray:
        """Ordonne 4 points pour former un rectangle cohérent"""
        rect = np.zeros((4, 2), dtype="float32")
        s = pts.sum(axis=1)
        rect[0] = pts[np.argmin(s)]  # haut-gauche
        rect[2] = pts[np.argmax(s)]  # bas-droit
        diff = np.diff(pts, axis=1)
        rect[1] = pts[np.argmin(diff)]  # haut-droit
        rect[3] = pts[np.argmax(diff)]  # bas-gauche
        return rect

    def detect_macbeth_in_scene(self, frame_raw: np.ndarray) -> tuple[np.ndarray, list[tuple[int, int, int, int]]]:
        """Détecte et analyse la charte Macbeth dans une image"""
        if frame_raw is None:
            raise ValueError("Image invalide")

        # Détection du cadre noir
        frame_hsv = cv2.cvtColor(frame_raw, cv2.COLOR_BGR2HSV)
        lower_black = np.array([0, 0, 0])
        upper_black = np.array([180, 100, 30])
        frame_black_mask = cv2.inRange(frame_hsv, lower_black, upper_black)

        # Nettoyage du masque
        morphology_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (7, 7))
        frame_black_mask = cv2.morphologyEx(frame_black_mask, cv2.MORPH_CLOSE, morphology_kernel)
        frame_black_mask = cv2.morphologyEx(frame_black_mask, cv2.MORPH_OPEN, morphology_kernel)

        # Détection du contour
        contours_black, _ = cv2.findContours(frame_black_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        frame_border_contour = None
        border_area_max = 0
        
        for contour in contours_black:
            contour_area = cv2.contourArea(contour)
            if contour_area > 1000:
                contour_perimeter = cv2.arcLength(contour, True)
                contour_approx = cv2.approxPolyDP(contour, 0.02 * contour_perimeter, True)
                if len(contour_approx) == 4 and contour_area > border_area_max:
                    border_area_max = contour_area
                    frame_border_contour = contour_approx

        if frame_border_contour is None:
            raise ValueError("Aucun grand rectangle noir détecté")

        # Transformation perspective
        corner_points = self._order_points(frame_border_contour.reshape(4, 2))
        
        # Correction du calcul des dimensions cibles
        width1 = np.linalg.norm(corner_points[1] - corner_points[0])
        width2 = np.linalg.norm(corner_points[2] - corner_points[3])
        target_width = int(float(max(width1, width2)))
        
        height1 = np.linalg.norm(corner_points[3] - corner_points[0])
        height2 = np.linalg.norm(corner_points[2] - corner_points[1])
        target_height = int(float(max(height1, height2)))

        dst_points = np.array([[0, 0],
                             [target_width - 1, 0],
                             [target_width - 1, target_height - 1],
                             [0, target_height - 1]], dtype=np.float32)
        
        perspective_matrix = cv2.getPerspectiveTransform(corner_points.astype(np.float32), dst_points)
        frame_warped = cv2.warpPerspective(frame_raw, perspective_matrix, (target_width, target_height))

        # Détection des carrés
        squares = self._detect_color_squares(frame_warped)
        
        return frame_warped, squares

    def _detect_color_squares(self, frame_warped):
        """Détecte les carrés de couleur dans l'image redressée"""
        frame_edges = cv2.Canny(cv2.cvtColor(frame_warped, cv2.COLOR_BGR2GRAY), 50, 200)
        contours, _ = cv2.findContours(frame_edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        
        squares = []
        for c in contours:
            peri = cv2.arcLength(c, True)
            approx = cv2.approxPolyDP(c, 0.02 * peri, True)
            if len(approx) == 4:
                x, y, w, h = cv2.boundingRect(approx)
                if w > 10 and h > 10:
                    squares.append((x + 5, y + 20, w - 10, h - 40))
        
        return self._organize_squares(squares)

    def _organize_squares(self, squares):
        """Organise les carrés détectés en grille 4x6"""
        squares.sort(key=lambda s: (s[1], s[0]))
        organized_squares = []
        
        rows = []
        current_row = []
        last_y = None
        
        for square in squares:
            if last_y is None or abs(square[1] - last_y) < 20:
                current_row.append(square)
            else:
                if current_row:
                    rows.append(sorted(current_row, key=lambda s: s[0]))
                current_row = [square]
            last_y = square[1]
            
        if current_row:
            rows.append(sorted(current_row, key=lambda s: s[0]))
            
        for row in rows:
            organized_squares.extend(row)
            
        return organized_squares

    def get_average_colors(self, frame_raw: np.ndarray, detect_squares: bool = True) -> list[tuple[int, int, int]]:
        """Calcule les couleurs moyennes des carrés"""
        frame_warped, squares = self.detect_macbeth_in_scene(frame_raw)
        
        colors = []
        for x, y, w, h in squares:
            roi = frame_warped[y:y+h, x:x+w]
            mean_color = cv2.mean(roi)[:3]
            colors.append(tuple(int(c) for c in mean_color))
            
        colors.reverse()  # Pour correspondre à l'ordre standard
        return colors